### DISCUSSION TITLE
Relative luminance proposal.

### DISCUSSION THREAD

**Discussion by setanarut** [godotengine/godot-proposals#4246] [Engine Core]
Replace the `Color.get_luminance()` formula with the WCAG standard relative luminance formula. Or provide an additional function called `get_relative_luminance()`. This is a formula used for the WCAG contrast test. It just needs to be linearized.
(Godot 4 Alpha 4)

```gdscript
# current Color.get_luminance() formula of Godot 4
func luminance(c: Color) -> float:
    return 0.2126 * c.r + 0.7152 * c.g + 0.0722 * c.b

# WCAG relative luminance formula
func relative_luminance(c: Color) -> float:
    c.r = linearize(c.r)
    c.g = linearize(c.g)
    c.b = linearize(c.b)
    return 0.2126 * c.r + 0.7152 * c.g + 0.0722 * c.b

func linearize(v: float) -> float:
    if v <= 0.03928:
        return v / 12.92
    else:
        return pow((v + 0.055) / 1.055, 2.4)
```

Color.get_luminance()
https://github.com/godotengine/godot/blob/c8e62555f6005cdec1f2172478724ab4dc6f93ad/core/math/color.h#L97

**Comment by clayjohn** [comment]
The WCAG uses the same formula as Godot for luminance ``0.2126f * r + 0.7152f * g + 0.0722f * b``.

The difference you are seeing there comes from the fact that the WCAG recommends that all content be specified with sRGB encoding, so it needs to be converted to linear before calculating the luminance (see the page on [relative luminance](https://www.w3.org/WAI/GL/wiki/Relative_luminance) and the page on [contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) which mentions using sRGB encoding vs other types of encoding).

Godot does not assume you are using an sRGB encoding. It assumes that you are using linear encoding when you are working with Colors directly. 

That being said, in Godot you can quickly convert sRGB encoded colours into linear space using ``to_linear()``. For example, if you have a color named ``col`` that is encoded in sRGB space you can get luminance with:
```
var luminance = col.to_linear().get_luminance()
```

Remember, there is a world of a difference between a game engine and the Web Content Accessibility Guidelines. A process that works for one may not work for the other. Importantly though, the WCAG does not recommend that linearization and luminance be packaged together in a single function. They say that _if the value is in sRGB space, then you need to linearize before applying this formula_.

**Reply by setanarut** [reply]
Maybe you're right `to_linear()` didn't exist in `Godot 4 alpha 4` and I hadn't thought of it. But on the other hand, non-perceptual contrast has no use case. So get_luminance() is completely unnecessary with its current formula. (contrast measurement will be inaccurate). That's why I wanted it to be in the core as C++.

**Comment by clayjohn** [comment]
> Maybe you're right to_linear() didn't exist in Godot 4 alpha 4 and I hadn't thought of it

I mentioned it as I use it often in the engine code, but looking closer now I can see that it is not exposed outside of the engine, which means my suggestion is pretty unhelpful!

> So get_luminance() is completely unnecessary with its current formula. (contrast measurement will be inaccurate)

This is only true if you assume that every Color is specified in sRGB space. If users are working in linear space already then ``get_luminance()`` is correct.

I think that the best resolution may be to expose ``to_linear()`` and ``to_srgb()`` and then document that ``get_luminance()`` relies on the colour being in linear space to return an accurate relative luminance value.

**Reply by Calinou** [reply]
https://github.com/godotengine/godot/pull/59456
